File: src/main/java/ru/citeck/idea/files/navigateItemsProviders/JsonJournalItemsProvider.java
package ru.citeck.idea.files.navigateItemsProviders;

import com.intellij.json.psi.JsonFile;
import com.intellij.psi.PsiFile;
import org.jetbrains.annotations.Nullable;

import ru.citeck.idea.files.FileTypeService;
import ru.citeck.idea.files.NavigateInFileItem;
import ru.citeck.idea.files.NavigateInFileItemsProvider;
import ru.citeck.idea.files.types.citeck.ui.Journal;
import ru.citeck.idea.utils.CiteckPsiUtils;

import java.util.Collection;
import java.util.Optional;
import java.util.stream.Collectors;

public class JsonJournalItemsProvider implements NavigateInFileItemsProvider {

    @Override
    public @Nullable Collection<NavigateInFileItem> getItems(PsiFile psiFile) {

        if (!(psiFile instanceof JsonFile)) {
            return null;
        }

        return Optional
            .ofNullable(FileTypeService.getInstance().getFileType(psiFile))
            .filter(Journal.JSON.class::isInstance)
            .map(Journal.JSON.class::cast)
            .map(journal -> Journal.JSON.getColumns(psiFile))
            .map(columns -> columns
                .stream()
                .map(column -> new NavigateInFileItem(CiteckPsiUtils.getProperty(column, "id"), column))
                .collect(Collectors.toList())
            )
            .orElse(null);
    }

}

File: src/main/kotlin/ru/citeck/idea/artifacts/action/navigate/providers/JsonJournalItemsProvider.kt
package ru.citeck.idea.artifacts.action.navigate.providers

import com.intellij.json.psi.JsonArray
import com.intellij.json.psi.JsonFile
import com.intellij.json.psi.JsonObject
import com.intellij.json.psi.JsonValue
import com.intellij.psi.PsiFile
import ru.citeck.idea.artifacts.ArtifactTypes
import ru.citeck.idea.utils.CiteckPsiUtils
import java.util.stream.Collectors

class JsonJournalItemsProvider : NavigateInFileItemsProvider {

    override fun getItems(psiFile: PsiFile?): List<NavigateInFileItem>? {

        if (psiFile !is JsonFile) {
            return null
        }

        if (!ArtifactTypes.isJournal(psiFile)) {
            return null
        }

        return getColumns(psiFile).map { column ->
            NavigateInFileItem(CiteckPsiUtils.getProperty(column, "id"), column)
        }
    }

    private fun getColumns(psiFile: PsiFile): List<JsonObject> {

        if (psiFile !is JsonFile) {
            return emptyList()
        }

        val root = psiFile.topLevelValue as? JsonObject ?: return listOf()

        val columns = root.findProperty("columns")
        if (columns == null || columns.value !is JsonArray) {
            return listOf()
        }

        return (columns.value as JsonArray)
            .valueList
            .stream()
            .filter { column: JsonValue? -> column is JsonObject }
            .map { column: JsonValue -> column as JsonObject }
            .collect(Collectors.toList())
    }
}
